contracts/SignerBondsUniswapV2.sol

```diff
@@ -0,0 +1,254 @@
+    ///        gas-expensive environment. Deadline equal to 0 will effectively
+    //         make swaps impossible.
+    function setSwapDeadline(uint256 _swapDeadline) external onlyOwner {
+        swapDeadline = _swapDeadline;
+    }
+
+    /// @notice Swaps signer bonds on Uniswap v2 exchange.
+    /// @dev Swaps the given ETH amount for the target token using the Uniswap
+    ///      decentralized exchange. The maximum ETH amount is capped by the
+    ///      contract balance. Some governance parameters are applied on the
+    ///      transaction. The swap's price impact must fit within the
+    ///      maximum allowed price impact and the transaction is constrained
+    ///      with the slippage tolerance and deadline. Acquired target tokens
+    ///      are sent to the recipient address set during contract construction.
+    /// @param amount Amount to swap.
+    function swapSignerBondsOnUniswapV2(uint256 amount) external {
+        require(amount > 0, "Amount must be greater than 0");
+        require(amount <= address(this).balance, "Amount exceeds balance");
+
+        // Setup the swap path. WETH must be the first component.
+        address[] memory path = new address[](2);
+        path[0] = uniswapRouter.WETH();
+        path[1] = targetToken;
+
+        // Calculate the maximum output token amount basing on pair reserves.
+        // This value will be used as the minimum amount of output tokens that
+        // must be received for the transaction not to revert.
+        // This value includes liquidity fee equal to 0.3%.
+        uint256 amountOutMin = uniswapRouter.getAmountsOut(amount, path)[1];
+
+        require(
+            isAllowedPriceImpact(amountOutMin),
+            "Price impact exceeds allowed limit"
+        );
+
+        // Include slippage tolerance into the minimum amount of output tokens.
+        amountOutMin = amountOutMin
+            .mul(BASIS_POINTS_DIVISOR.sub(slippageTolerance))
+            .div(BASIS_POINTS_DIVISOR);
+
+        // slither-disable-next-line arbitrary-send,reentrancy-events
+        uint256[] memory amounts =
+            uniswapRouter.swapExactETHForTokens{value: amount}(
+                amountOutMin,
+                path,
+                targetTokenRecipient,
+                /* solhint-disable-next-line not-rely-on-time */
+                block.timestamp.add(swapDeadline)
+            );
+
+        emit UniswapV2SwapExecuted(amounts);
+    }
+
+    /// @notice Checks the price impact of buying a given amount of tokens
+    ///         against the maximum allowed price impact limit.
+    /// @param amount Amount of tokens.
+    /// @return True if the price impact is allowed, false otherwise.
+    function isAllowedPriceImpact(uint256 amount) internal view returns (bool) {
+        // Get reserve of the target token.
+        address WETH = uniswapRouter.WETH();
+        address token0 = WETH < targetToken ? WETH : targetToken;
```

Comment:
I'm a bit confused why comparing addresses will tell us which one is which. We can also add `token0()` to `IUniswapV2Pair` interface and get the address of `token0`. Then  `uint256 targetTokenReserve = WETH == token0 ? reserve1 : reserve0;` remains as is

Reply:
This code is based on https://github.com/Uniswap/uniswap-v2-periphery/blob/master/contracts/libraries/UniswapV2Library.sol. I think that comparing addressed using `<` is cheaper than calling a contract method.

Reply:
Ah, I see now, when they [create](https://github.com/Uniswap/uniswap-v2-core/blob/master/contracts/UniswapV2Factory.sol#L23) a pair the tokens are sorted. Okay, I didn't realize that. Sure, let's leave it then.